### PR TITLE
Keep import files cached on s3 and local disk and allow import from there

### PR DESCRIPTION
**Pull request by aldavidson**
so that we can still deploy a new system when Ordnance Survey have removed AddressBase from our FTP site, which they do 21 days after ordering (even though we won't get a new file until the next update - which might be another 2 months away)

### CODE REVIEW

File: postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+from django.conf import settings
+
+from boto.s3.connection import S3Connection
+from boto.s3.key import Key
+
+class S3Adapter(object):
+
+    def __init__(self, connection=None, bucket=None):
+        self.connection = connection or self.connect()
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py]: It's more Pythonic to test for `None` with the `is` operator, eg:
```python
self.connection = connection
if connection is None:
    self.connection = self.connect()
```

File: postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        self.bucket = bucket or self.bucket()
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py]: assigning to `self.bucket` shadows the `self.bucket()` method.

File: postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import boto
+import hashlib
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py]: unused import

File: postcodeinfo/apps/postcode_api/downloaders/postcode_gss_code_downloader.py [hunk 4]
```diff
@@ -11,16 +11,23 @@ def download(self, target_dir='/tmp/', force=False):
         most_recent_file_url = self._target_href()
 
         dl_mgr = DownloadManager()
-        return dl_mgr.download_if_needed(most_recent_file_url,
+        #import pdb; pdb.set_trace()
+
+        return dl_mgr.retrieve(most_recent_file_url,
                                          target_dir,
                                          force)
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/postcode_gss_code_downloader.py]: This change has broken the arbitrary indentation of the lines 17 and 18. Better to format this like:
```python
return dl_mgr.retrieve(
    most_recent_file_url,
    target_dir,
    force)
```

File: postcodeinfo/apps/postcode_api/downloaders/local_authorities_downloader.py [hunk 5]
```diff
@@ -9,7 +9,7 @@ def download(self, target_dir='/tmp/', force=False):
 
         dl_mgr = DownloadManager()
 
-        return dl_mgr.download_if_needed(most_recent_file_url,
+        return dl_mgr.retrieve(most_recent_file_url,
                                          target_dir,
                                          force)
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/local_authorities_downloader.py]: indentation doesn't match

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 6]
```diff
@@ -6,27 +6,82 @@
...
 
-    def do_download(self, url, dirpath, headers):
-        local_path = self.download_to_dir(url, dirpath, headers)
-        download_record = self.record_download(url, dirpath, headers)
-        return download_record.local_filepath
+    def _in_local_storage(self, local_path):
+        return os.path.exists(local_path)
+
+    def up_to_date(self, copy_timestamp, source_timestamp):
+        return (copy_timestamp >= source_timestamp)
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Parens not needed

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 7]
```diff
@@ -6,27 +6,82 @@
...
+
+    def s3_object_up_to_date(self, s3_object, remote_timestamp):
+        return ((s3_object != None)
+                and self.up_to_date(self.format_time_for_orm(
+                                        s3_object.last_modified),
+                                    remote_timestamp))
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Don't use `==` or `!=` to compare to None. This should be `is not None`. Also, this would be more Pythonic as
```python
if s3_object is not None:
    return self.up_to_date(self.format_time_for_orm(s3_object.last_modified), remote_timestamp)
```
If `s3_object` is `None` the method will return `None` by default

Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Or maybe you want an explicit `return False` if this is always supposed to return a boolean

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 8]
```diff
@@ -6,27 +6,82 @@
...
-        else:
-            print 'no download needed'
-            return None
+        return self.format_time_for_orm(headers['last-modified'])
+
+    def local_copy_up_to_date(self, local_file_path, remote_timestamp):
+        result = (self._in_local_storage(local_file_path)
+                  and self.up_to_date(self.format_time_for_orm(
+                      os.path.getmtime(local_file_path)),
+            remote_timestamp))
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Indentation looks weird here (although that might just be Github?). I think this could be more clearly written as:
```python
local_copy = self._in_local_storage(local_file_path)
up_to_date = self.up_to_date(self.format_time_for_orm(os.path.getmtime(local_file_path), remote_timestamp)
result = local_copy and up_to_date
```

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py [hunk 9]
```diff
@@ -13,11 +16,28 @@ def download(self, local_dir='/tmp/', force=False):
...
-        return self.download_manager.download_all_if_needed('./*_csv.zip',
-                                                            local_dir,
-                                                            force)
+        try:
+            self.ftp_client().cwd(self._source_dir())
+            return self.download_manager.download_all_if_needed('./*_csv.zip',
+                                                                local_dir,
+                                                                force)
+        except ftplib.error_perm:
+            print 'FTP error - looking for local files'
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: These print lines should be `logging.debug` or `logging.info` calls instead.

Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: Or, `logging.warning` in this case

File: postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py (hunk 1 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py (hunk 3 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/s3_adapter.py]: fixed

File: postcodeinfo/apps/postcode_api/downloaders/postcode_gss_code_downloader.py (hunk 4 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/postcode_gss_code_downloader.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/local_authorities_downloader.py (hunk 5 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/local_authorities_downloader.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 6 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 7 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: ok, tidied that method up to be clearer (the explicit False was needed)

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 8 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: tidied this block up now

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py (hunk 9 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py [hunk 10]
```diff
@@ -13,23 +17,40 @@ def download(self, local_dir='/tmp/', force=False):
...
+            # Ordnance Survey very kindly remove the files from our
+            # FTP area 21 days after they are first put there, despite
+            # there not being an update available for maybe another 2 months.
+            # So if we try to cd to our order dir
+            # and it's not there, we'll end up here.
+            # So, we then look for a load of csv.zip files in local_dir
+            # If there are csv.zip files locally, return them
+            # If not, look on s3.
+            # If they're on s3, download them and return the local paths.
+            # Otherwise FAIL
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: I think this comment would be better as a document string on the method/class.

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py [hunk 11]
```diff
@@ -13,23 +17,40 @@ def download(self, local_dir='/tmp/', force=False):
...
-        self.download_manager.ftp_client.cwd(self._source_dir())
 
-        return self.download_manager.download_all_if_needed('./*_csv.zip',
-                                                            local_dir,
-                                                            force)
+        try:
+            self.ftp_client().cwd(self._source_dir())
+            return self.download_manager.download_all_if_needed('./*_csv.zip',
+                                                                local_dir,
+                                                                force)
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: Arbitrary indentation

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py [hunk 12]
```diff
@@ -13,23 +17,40 @@ def download(self, local_dir='/tmp/', force=False):
...
+            return self.local_files(local_dir) or self.files_from_s3(local_dir)
+
+    def ftp_client(self):
+        return self.download_manager.ftp_client
 
     def _username(self):
         username = os.environ.get('OS_FTP_USERNAME')
         if not username:
-            print 'OS_FTP_USERNAME not set!'
+            logging.warning('OS_FTP_USERNAME not set!')
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: For this and OS_FTP_PASSWORD, do we want to raise an exception instead?

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py [hunk 13]
```diff
@@ -37,3 +58,19 @@ def _source_dir(self):
         # TODO - this is currently just the order number.
         # Make this dynamic based on the most recent directory?
         return os.environ.get('OS_FTP_ORDER_DIR', '../from-os/DCS0001654526/')
+
+    def local_files(self, local_dir):
+        files = glob.glob(os.path.join(local_dir, '*csv.zip'))
+        logging.debug('found {num_files} files in {path} matching {pattern}'.format(num_files=len(files), path=local_dir, pattern='*csv.zip'))
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: Line > 80 chars. Is this something we care about?

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py [hunk 14]
```diff
@@ -37,3 +58,19 @@ def _source_dir(self):
...
+        return files
+
+    def files_from_s3(self, local_dir):
+        files = []
+        s3 = S3Adapter()
+        s3_files = s3.bucket.list('AddressBase')
+        logging.debug( 'found %i files in s3 matching %s' % (len(s3_files), 'AddressBase') )
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: > 80 chars

Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: Extra spaces around the arguments to `debug`

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py [hunk 15]
```diff
@@ -37,3 +58,19 @@ def _source_dir(self):
...
+        for key in s3_files:
+            logging.debug( 'downloading %s' % key.name )
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: Extra spaces

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 16]
```diff
@@ -6,31 +7,91 @@
...
 
-    def download_if_needed(self, url, dirpath, force=False):
+    # Reasons for this workflow:
+    # 1) Ordnance Survey only gives us AddressBase files
+    #    for a limited time after updates are released (21 days)
+    #    then it removes the file from their FTP server
+    #
+    # 2) Our release process re-creates an entirely new
+    #    docker container each time, so we can't just
+    #    cache the files locally
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: This comment would be better as a docstring, eg:
```python
"""Reasons for this workflow:
1. Ordnance Survey ...
2. Our release process ...
"""
```

Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Also, if a docstring, it should start on the first line inside the `retrieve` method (if that's what it is commenting on).

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 17]
```diff
@@ -6,31 +7,91 @@
...
+    def retrieve(self, url, local_dir_path, force=False):
+        remote_timestamp = self.get_last_modified(url)
+        local_file_path = self.filename(local_dir_path, url)
+
+        # if we have an up-to-date local copy, there's nothing to do
+        if not self.local_copy_up_to_date(local_file_path,
+                                          remote_timestamp):
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: This line doesn't need to be split.

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 18]
```diff
@@ -6,31 +7,91 @@
...
+    def get_from_s3(self, local_file_path, url, remote_timestamp):
+        # if we *don't* have an up-to-date local copy
+        # do we have one on s3?
+        s3 = self.s3_adapter()
+        s3_key = s3.key(url)
+        s3_object = s3.file(s3_key)
+
+        if self.s3_object_is_up_to_date(s3_object, remote_timestamp):
+            # if so, mirror it locally
+            logging.info('downloading from s3 to {path}'.format(path=local_file_path))
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Line > 80 chars

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 19]
```diff
@@ -6,31 +7,91 @@
...
+            s3.download(s3_object, local_file_path)
+        else:
+            # so let's grab a local copy
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: This comment seems to be orphaned

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 20]
```diff
@@ -6,31 +7,91 @@
...
+            logging.info('downloading from {url} to {path}'.format(url=url, path=local_file_path))
+            # nope, s3 object either not there or out of date
+            self.download_to_file(url, local_file_path)
+            # and upload *that* to s3 for later use
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: These comments don't add much value - they repeat what the code says.

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 21]
```diff
@@ -6,31 +7,91 @@
...
+        return self.format_time_for_orm(headers['last-modified'])
+
+    def local_copy_up_to_date(self, local_file_path, remote_timestamp):
+        local_copy = self._in_local_storage(local_file_path)
+        if local_copy:
+            local_mod_time = os.path.getmtime(local_file_path)
+            formatted_mod_time = self.format_time_for_orm(
+                                    local_mod_time)
+            result = self.up_to_date(formatted_mod_time,
+                                     remote_timestamp)
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: These lines don't need to be split.

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 22]
```diff
@@ -6,31 +7,91 @@
...
+            return result
         else:
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Don't need an `else` if the last line in the `if` branch is return, eg:
```python
if cond:
    return result
return False
```

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 23]
```diff
@@ -6,31 +7,91 @@
...
-        local_path = self.download_to_dir(url, dirpath, headers)
-        download_record = self.record_download(url, dirpath, headers)
-        return download_record.local_filepath
+    def s3_object_is_up_to_date(self, s3_object, remote_timestamp):
+        if s3_object is None:
+            return False
+        else:
+            mod_time_on_s3 = self.format_time_for_orm(s3_object.last_modified)
+            return self.up_to_date(mod_time_on_s3,
+                                   remote_timestamp)
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: This line doesn't need to be split.

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 24]
```diff
@@ -45,12 +106,12 @@ def download_to_file(self,
             count = 0
             for chunk in r.iter_content(chunk_size):
                 if content_length and count % 100 == 0:
-                    print '{0} bytes of {1}'.format(count*chunk_size,
-                                                    content_length)
+                    logging.debug('{0} bytes of {1}'.format(count*chunk_size,
+                                                            content_length))
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Arbitrary indentation. IMO, it's better to do:
```python
logging.debug('{0} bytes of {1}'.format(
    count * chunk_size,
    content_length))
```

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py [hunk 25]
```diff
@@ -6,31 +7,91 @@
...
+        return copy_timestamp >= source_timestamp
 
-    def do_download(self, url, dirpath, headers):
-        local_path = self.download_to_dir(url, dirpath, headers)
-        download_record = self.record_download(url, dirpath, headers)
-        return download_record.local_filepath
+    def s3_object_is_up_to_date(self, s3_object, remote_timestamp):
+        if s3_object is None:
+            return False
+        else:
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: Don't need an `else` here. Also, I think the logic should be swapped, eg:
```python
if s3_object is not None:
    # do stuff
    return # something
return False
```

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py (hunk 10 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py (hunk 11 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: not any more!

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py (hunk 14 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: fixed

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py (hunk 15 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: fixed

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 16 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 17 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: fixed

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 18 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: fixed

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 19 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: re-organised comments

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 20 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: ....probably because I wrote the comments first, before any code :)

Agreed, I've taken them out now

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 22 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: done

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 23 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: fixed

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 24 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: fixed

File: postcodeinfo/apps/postcode_api/downloaders/download_manager.py (hunk 25 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/download_manager.py]: fixed

**Comment by aldavidson** [comment]
ok, @andyhd - ready for round 3

File: postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py (hunk 12 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/downloaders/addressbase_basic_downloader.py]: not sure - happy to discuss

**Comment by andyhd** [comment]
LGTM
